Урок 08.  Компоненты. Взаимодействие между модулями. Общий модуль

rusdec opened the pull request:
- Добавлен SharedModule;
- Добавлен компонент app-button;
- Добавлена крутяшка app-spinner;
- Добавлен eslint;
- Код отформатирован согласно правилам eslint;

rusdec committed:
[lesson-07] Добавлен SharedModule;
Добавлен компонент app-button;
Добавлена крутяшка app-spinner;
Добавлен eslint;
Отформатирован код согласно правилам eslint;

rusdec committed:
[lesson06] Переминование text в title;
Вынос тега li в родительский компонент;
Обрезание длинного текста заголовка.

viktor-berezin commented:
Привет, заапрувил предыдущий PR, так как там минорные комменты.
Пожалуйста пересоздай этот PR с направлением из lesson-7 в main. (тебе возможно потребуется находясь в lesson-7 надо сделать merge из main)

Я все комменты перенесу в новый PR. Или можешь пофиксить их сразу.

viktor-berezin commented on src/app/components/todo-list-item-new/todo-list-item-new.component.ts:
```diff
@@ -1,16 +1,17 @@
-  @Output('item-created') itemCreated = new EventEmitter<string>();
+  @Output() itemCreated = new EventEmitter<string>();
 
   isValid(): boolean {
```
пожалуйста удали метод isInvalid() (это коммент из прошлого PR)

лучше не плодить лишние методы, компонент и без этого имеет свойство расти... тут точно нужно обойтись без метода.

просто там где требуется используй !isValid()

rusdec replied:
Поправил

viktor-berezin commented on src/app/components/todo-list/todo-list.component.html:
```diff
@@ -4,13 +4,14 @@
   <mat-card-content>
-    <mat-list>
+    <app-spinner *ngIf="isLoading"></app-spinner>
+    <mat-list *ngIf="!isLoading">
```
на этом этапе мы уже используем:
`*ngIf="!isLoading; else loadingSpinner"`

<app-spinner #loadingSpinner></app-spinner> располагаем внизу шаблона

rusdec replied:
Поправил.
А есть ли какое-то правило, какой именно тег нужно располагать внизу страницы в ng-template, а в каком писать условие и располагать в общем html-коде?
Я бы, например, избегая инверсии в условии, оставил бы индикатор загрузки (spinner) в общем блоке html-кода, а mat-list вынес бы вниз шаблона в ng-template:

```
<app-spinner *ngIf="isLoading; else items"></app-spinner>
...
<ng-template #items>
  <mat-list>
    ...
  </mat-list>
</ng-template>
```

viktor-berezin replied:
слушай да мне тоже иногда хочется первым написать условие без !, но правила нет, и по идее можно.
но всё же принято сначала основной контент писать, а всякие спиннеры или сообщения о том что элементов в списке нет располагать внизу шаблона и в итоге они становятся по ;else и условие со знаком ! в этом случае.

viktor-berezin commented on src/app/modules/shared/components/button/button.component.scss:
```diff
@@ -0,0 +1,34 @@
+:host {
+    .app-button {
```
давай удалим класс app-button и определим стили через тег
```
button {
...
}
```

rusdec replied:
Поправил

viktor-berezin commented on src/app/modules/shared/components/button/button.component.scss:
```diff
@@ -0,0 +1,34 @@
+            opacity: 0.8;
+        }
+    
+        &--primary {
```
&.primary {..}
&.warning {...}

rusdec replied:
Поправил

viktor-berezin commented on src/app/modules/shared/components/button/button.component.html:
```diff
@@ -0,0 +1,3 @@
+<button [class]="cssClass()" [disabled]="disabled">
```
если удалить css класс app-button (он не требуется, у нас только одна кнопка) то можно будет просто написать [class]="color" и удалить метод cssClass() совсем.
меньше кода - проще компонента, а функциональность такая же и читать удобнее

rusdec replied:
Поправил

rusdec committed:
[lesson-07] Правки по комментариям:
индикатор загрузки вынесен в ng-template;
стили кнопок определяются для html-элемента button;
использование метода TodoListItemNew#isValid() вместо TodoListItemNew#isInvalid().
